Multiprocessing

nasko25 opened the pull request:
- Implemented a context switcher and a dispatcher.
- Rewrote the software and irq handlers (they seems to work, but further testing would be beneficial)

fayalalebrun committed:
Mark memory slices as reserved to prevent them being freed

fayalalebrun committed:
Use standard bash random variable for random number generation

fayalalebrun committed:
Remove outdated comment

jdonszelmann committed:
Merge pull request #60 from ValentijnvdBeek/make-file-program-checking

Add a docker image and detect local toolchains

fayalalebrun committed:
Merge remote-tracking branch 'origin/master' into random-tests

jdonszelmann committed:
Merge pull request #63 from fayalalebrun/random-tests

Seed-based randomized test order

jdonszelmann committed:
fix cmakelist

fayalalebrun committed:
Use docker in CI; Don't echo random number if pipe is closed

fayalalebrun committed:
Separate build and test stages

fayalalebrun committed:
Merge pull request #70 from fayalalebrun/random-tests

Use docker in CI; Don't echo random number if pipe is closed

jdonszelmann committed:
Merge branch 'master' of github.com:rellermeyer/course_os into fix_clion

jdonszelmann committed:
Merge pull request #64 from rellermeyer/fix_clion

fix cmakelist

Commit:
Skrewing around with context switching (DOES NOT WORK, I THINK)

Commit:
Improved version of context switching, might actually work?

Commit:
Improved code readability

Commit:
Improved comments

nasko25 committed:
Started adding a skeleton of a syscall helper that will hopefully be used by the processes (when we implement them).

For now it does not do anything (I am trying to make it work with the already implemented software interrupt handlers by calling SWI from assembly, because it looks like this was intended by the other developers).

nasko25 committed:
Found a bug in interrupt.c while testing.

Also for some reason calling code from syscall.c causes kernel panic, but running the same code directly does not (not entering the switch statement in syscall() does not cause the kernel panic, so it should be a problem with the inline assembly).

nasko25 committed:
Fixed the bug in interrupt.c

There was a problem in the way inline assembly was executed, so I changed how the variables are taken from the registers.

I am not really sure if it is a great way, because I read the syntax might be compiler-dependent, but using registers is already pretty system-dependent, so it shouldn't matter.

Also tested more the syscalls code, and it still behaves weirdly, so I made the comment regarding the issue more descriptive.

fayalalebrun committed:
Merge remote-tracking branch 'origin/master' into memdetect

fayalalebrun committed:
Merge remote-tracking branch 'origin/master' into memdetect

fayalalebrun committed:
Add POSIX error handling; add test for invalid slice

fayalalebrun committed:
Change wording of test to "reserved"

nasko25 committed:
Very weird. Adding an empty kprintf call to the end of the dummy case in the switch statement fixes the kernel panic.

This needs to be investigated further. Probably some registers are incorrectly set before the software interrupt because of the inline assembly?

Dimitar2000 committed:
Fix small bug in interrupt.c

Commit:
Initial very minimal interface for PCB and scheduler

nasko25 committed:
Fixed the weird kernel panic.

For some reason the software interrupt handler changed lr, so after returnting from the SWI call, the syscall() function returned to where lr was pointing to (somewhere in software_interrupt_handler()), which caused a kernel panic at some point, probably because of dereferencing an invalid pointer.

Commit:
expanded the scheduler and PCB interface + added implementations

Commit:
added TODO, how to handle empty queue?

Commit:
fixed previous TODO

Commit:
Screwing around with things, doesn't work

Commit:
Resolve merge conflicts

Commit:
Resolve errors

fayalalebrun committed:
Merge pull request #61 from fayalalebrun/memdetect

Enable memory detection for PMM; Fix wrong amount of memory mapped for peripherals

ValentijnvdBeek committed:
Find native GDB & stdbool C99 compliance

ValentijnvdBeek committed:
Add memchr and memcmp

ValentijnvdBeek committed:
Add support for printing pointers in kprintf

ValentijnvdBeek committed:
Implement most of string.h

ValentijnvdBeek committed:
Add a tokenizer method and push_to_back for VPSSL

ValentijnvdBeek committed:
string.h tests

ValentijnvdBeek committed:
Documentation changes

nasko25 committed:
Give an int pointer to the SWI handler when calling the fork syscall.

Now the software interrupt handler can return the pid of the newly created process.

nasko25 committed:
Added build/ and .cache/ to the gitignore.

(I think they were generated by vscode)

Commit:
Almost working context saver, and fix gitignore issues

Commit:
Added more docs, and todos

Commit:
Fix syntax error

Commit:
Improves vscode launch config to work with default cpp extension, and launches debugger in one task

Commit:
Removes non-gitignored build files

Commit:
Merge improved vscode launch file

Commit:
Removes vscode launch files from gitignore

Commit:
Adds root build directory to gitignore

Commit:
minor change to gitignore and better comment

Commit:
Merge remote-tracking branch 'origin/scheduler_setup_zohar' into scheduler_setup_Erik

fayalalebrun committed:
Add Emacs gdb command variable set automatically

fayalalebrun committed:
Update readme related to Emacs GDB command

Commit:
fixed  a laundry list of errors in my original code

haven't used C in a while...

nasko25 committed:
Fixed indentation issues.

For some reason there were some tabs instead of spaces in .

nasko25 committed:
Pass arguments when calling SYS_kill.

The interfaces for SYS_fork and SYS_kill should be done.

nasko25 committed:
The syscall interface should be done.

All syscalls correctly pass parameters to the software interrupt handler.

Commit:
Saving context now works
Co-authored-by: Dimitar2000

Commit:
Better code formatting?

Commit:
Fix typo

jjq:q

Commit:
Correct docs

Commit:
Restructure struct to match loading order, added comments on switching vas

Commit:
TMP 2 :)

nasko25 committed:
Merged with scheduler_setup_zohar.

nasko25 committed:
Temp

Commit:
Trying to set up automation for path of gdb

nasko25 committed:
everything is kinda fine but r3 is overwritten

Commit:
Fixes debugging not working on open source vscode, and makes it more automated

Commit:
Replaces environment variable with working path

Commit:
Merge remote-tracking branch 'origin/scheduler_setup_zohar' into scheduler_setup_Erik

Commit:
"combined" execution states

Commit:
preemptively fixed mistakes, added kernelpanic if no processes active

Dimitar2000 committed:
Add initial documentation file.

Dimitar2000 committed:
Add initial documentation file.

ValentijnvdBeek committed:
Generic UART interface; OOP Uart Object; Reading characters; Cat program

ValentijnvdBeek committed:
New kernel panic graphic; First command

ValentijnvdBeek committed:
Merge remote-tracking branch 'labrun/emacs-gdb' into libc-stdio

Dimitar2000 committed:
Add header files for loader and elf.

Dimitar2000 committed:
Add header files for loader and elf.

Dimitar2000 committed:
Add a header file for expected common elf features

Dimitar2000 committed:
Add a header file for expected common elf features

Dimitar2000 committed:
Add structure definition and function declarations.

Dimitar2000 committed:
Add structure definition and function declarations.

Dimitar2000 committed:
Update elf_new.h: Add more declarations

Dimitar2000 committed:
Update elf_new.h: Add more declarations

Dimitar2000 committed:
ELF header parsing: Version 1

Dimitar2000 committed:
ELF header parsing: Version 1

Dimitar2000 committed:
Add declaration to loader.h

Dimitar2000 committed:
Add declaration to loader.h

nasko25 committed:
trying to switch to usermode; nothing works so far

the main problem right now is that the linker does not properly link to the virtual user address space, so no one can read from it

nasko25 committed:
now it should compile

I had modified the linker script and the compiler complained

ValentijnvdBeek committed:
WIP

Commit:
BOOIIIISSS

Commit:
Resolve merge conflicts

nasko25 committed:
merged with syscalls to get the allocate_page() fix

also memcpy the switch_to_usermode code to the allocated page but for some reason cannot jump there yet

nasko25 committed:
comments; switching to usermode appears to work

Commit:
Merge pull request #75 from rellermeyer/improved-vscode-config

Improves VSCode Launch Configuration

Commit:
Another WIP

nasko25 committed:
switching to userspace and calling swi finally works

nasko25 committed:
cleaned the code a bit

nasko25 committed:
trying to set a usermode stack

nasko25 committed:
context switching with software interrupts works

nasko25 committed:
passing values to the syscall handler works

nasko25 committed:
calling two consecutive syscalls works!!!

Dimitar2000 committed:
Update documentation. Add steps in overview.

Dimitar2000 committed:
Update documentation. Add steps in overview.

Dimitar2000 committed:
Change methods' signatures

Dimitar2000 committed:
Change methods' signatures

Dimitar2000 committed:
Change functions' signatures.

Dimitar2000 committed:
Change functions' signatures.

Dimitar2000 committed:
Add initial loader implementation code

Dimitar2000 committed:
Add initial loader implementation code

Dimitar2000 committed:
Fix errors and remove unnecessary files

Dimitar2000 committed:
Fix errors and remove unnecessary files

Commit:
Updates syscalls.c to work with context switching

Commit:
Update documentation on context switching

ValentijnvdBeek committed:
Improve GDB emacs support

nasko25 committed:
save the sp in r4 before calling svc

Dimitar2000 committed:
Fix CMake compiler and linker detection.

nasko25 committed:
made _save_state and _load_state macros

Dimitar2000 committed:
Add new elf enums and a function.

Dimitar2000 committed:
Add new elf enums and a function.

Dimitar2000 committed:
Merge pull request #79 from rellermeyer/fix_cmake

Fix CMake compiler and linker detection.

Dimitar2000 committed:
Loader: Implement program header table traversal and page allocation V1.

Dimitar2000 committed:
Loader: Implement program header table traversal and page allocation V1.

nasko25 committed:
saving and loading state restores the user's registers when returning from the software interrupt handler

fayalalebrun committed:
Add dtb files

fayalalebrun committed:
Add board option to makefile; Add linking of dtb into binary

fayalalebrun committed:
Add board configuration documentation

fayalalebrun committed:
Move dtb files to kernel directory

Commit:
Merge remote-tracking branch 'origin/scheduler_setup_Erik' into syscalls

Otherwise shit breaks

Commit:
Started trying to schedule a process

ValentijnvdBeek committed:
Stack protection, debug shell, more io functions

ValentijnvdBeek committed:
Add a debug API & arguments to commands

ValentijnvdBeek committed:
Toggable memory tracing

fayalalebrun committed:
Add raspberry pi zero dumped device tree

fayalalebrun committed:
Fix raspi0 core being disabled

fayalalebrun committed:
Fix timer tests failing

fayalalebrun committed:
Remove old memory detection mechanism; Add use of embedded dtb

fayalalebrun committed:
Add dtb struct; Add check for dtb magic number

Commit:
Context switching, now without user sp as parameter!

nasko25 committed:
initilizing and running two processes in parallel!

Commit:
Now also saves r0, updated and added more comments

Commit:
Fix typo

Commit:
Removes rogue folder

Commit:
Merge multiprocessing and libc-stdio

Commit:
Restructure assembly subroutines

Only include macro definitions

Commit:
Typo in include

fayalalebrun committed:
Add first implementation of dtb parser

ValentijnvdBeek committed:
Document debug

fayalalebrun committed:
Fix bugs in dtb parser, now functional

fayalalebrun committed:
Add dtb memory detection; Overhaul hardware detection using DTB

fayalalebrun committed:
Merge branch 'multiprocessing' into dtb

Dimitar2000 committed:
Loader: Add code for allocating stack and heap

Dimitar2000 committed:
Loader: Add code for allocating stack and heap

Commit:
Quick master update

Dimitar2000 committed:
Loader: Add comments.

Dimitar2000 committed:
Loader: Add comments.

Dimitar2000 committed:
Loader: Update code to use PCB struct.

Dimitar2000 committed:
Loader: Update code to use PCB struct.

Dimitar2000 committed:
Loader: Remove files from source set for CMake.

Dimitar2000 committed:
Loader: Remove files from source set for CMake.

fayalalebrun committed:
Add first implementation BCM2835 support

fayalalebrun committed:
Fix wrong address used for interrupt registers

fayalalebrun committed:
Refactor timer test to use chipset functions

Dimitar2000 committed:
Loader: Finish loader routine by initiating the exec state.

Dimitar2000 committed:
Loader: Finish loader routine by initiating the exec state.

Dimitar2000 committed:
Loader: Add comments to header file.

Dimitar2000 committed:
Loader: Add comments to header file.

Dimitar2000 committed:
Loader: Add comments to implementation file.

Dimitar2000 committed:
Loader: Add comments to implementation file.

Dimitar2000 committed:
Loader: Comment out init_state code.

Dimitar2000 committed:
Loader: Comment out init_state code.

fayalalebrun committed:
Fix system timer implementation; Fix failing test on raspi0

Dimitar2000 committed:
Loader: Add a simple assembly function for using an external macro.

Dimitar2000 committed:
Loader: Add a simple assembly function for using an external macro.

Commit:
Fix r7 not being restored properly

Commit:
Add pseudo code (not really but you get the point) for saving, and loading state in irq mode

ValentijnvdBeek committed:
Cleanup useless files

ValentijnvdBeek committed:
Fix CI

Commit:
Remove pseudocode, and add IRQ mode constant

Commit:
Add context switching in irq, and scheduler time-slices (UNTESTED...)

fayalalebrun committed:
Remove unused DTB files

fayalalebrun committed:
Add dtb readme

fayalalebrun committed:
Remove unnecessary memory variable

fayalalebrun committed:
Refactore dtb source to use existing ALIGN macro

fayalalebrun committed:
Move memory detection responsibility to hardware info

nasko25 committed:
debugging :)

nasko25 committed:
move the irq handler to assembly

ValentijnvdBeek committed:
Merge remote-tracking branch 'origin/multiprocessing' into cleanup

nasko25 committed:
irq interrupts work

Dimitar2000 committed:
Add an exec file into the kernel image file.

nasko25 committed:
added comments and cleaned the code a bit

ValentijnvdBeek committed:
WIP: Syscall method

Dimitar2000 committed:
Inject executable into kernel image file.

ValentijnvdBeek committed:
Merge branch 'multiprocessing' into cleanup

nasko25 committed:
bugfix

one register was left out when saving the state during an irq interrupt

ValentijnvdBeek committed:
Merge branch 'cleanup' into testing-framework-rewrite

ValentijnvdBeek committed:
Rewrite asserts and add a few new ones

ValentijnvdBeek committed:
Add an ANSI escape code library

ValentijnvdBeek committed:
Rewrite unit tests; Configurable memory tracing; Memory leak detection fix

ValentijnvdBeek committed:
Rewrite the test suite in Python and Jinja2

ValentijnvdBeek committed:
Use a Jinja macro for TEST_CREATE instead of a C macro

ValentijnvdBeek committed:
Rewrite the C macro into a Python macro

Dimitar2000 committed:
Add loader procedure to startup code. Fix some issues

Dimitar2000 committed:
Loader: Add INFO messages and fix bugs.

ValentijnvdBeek committed:
Organise test into test suites based on files

nasko25 committed:
fixed the failing tests

the order in which load_state_irq() restored the registers was wrong

ValentijnvdBeek committed:
Initial compile, doesn't run

ValentijnvdBeek committed:
Merge memallocator

ValentijnvdBeek committed:
Stylistic changes

ValentijnvdBeek committed:
Fix some typos

ValentijnvdBeek committed:
Update generate_tests.sh

Dimitar2000 committed:
Add a mkprex.sh

Dimitar2000 committed:
Loader: Change location of test executable in the linker script.

Dimitar2000 committed:
Loader: Fix a bug. Change return type of functions

Dimitar2000 committed:
Loader: Change elf_new.h to elf.h.

Dimitar2000 committed:
Loader: Fix macro usage. Change function subroutine.

ValentijnvdBeek committed:
Use R&L for good, and not evil.

ValentijnvdBeek committed:
Fix bug where the singly linked list assumed it ended with a NULL

ValentijnvdBeek committed:
Initial working version

Dimitar2000 committed:
Loader: Update loader functions.

Dimitar2000 committed:
Loader: Update shell script for making processes.

Dimitar2000 committed:
Loader: Rename docs file.

ValentijnvdBeek committed:
Implement other targets; Update Docker; Fix Github Action

Dimitar2000 committed:
Loader: Move test files to a new directory. Create a separate file for utility functions.

Dimitar2000 committed:
Remove sys and sched folder and their usages.

Dimitar2000 committed:
Loader: Change structure of files. Add defines.

Dimitar2000 committed:
Loader: Change structure of functions.

Dimitar2000 committed:
Loader: Update documentation.

Dimitar2000 committed:
Loader: Update documentation 2.

Dimitar2000 committed:
Loader: Update documentation 3.

ValentijnvdBeek commented:
The day has finally come for the merging storm! I will take a look it at a bit later today after my meetings

Commit:
Update kernel README.md

Updates the kernel readme to include the current state of the project, giving people starting on the project a better overview of what already is available to them.

fayalalebrun committed:
Merge remote-tracking branch 'origin/multiprocessing' into dtb

fayalalebrun committed:
Merge pull request #81 from fayalalebrun/dtb

Pi Zero Compatibility; DTB parser

nasko25 committed:
Updated the kernel readme

ValentijnvdBeek committed:
Merge pull request #83 from rellermeyer/testing-framework-rewrite

Testing framework rewrite

ValentijnvdBeek committed:
Merge branch 'allocator-merge' into cleanup

ValentijnvdBeek committed:
Merge remote-tracking branch 'origin/elf_loader' into cleanup

ValentijnvdBeek committed:
Fix merge bugs

ValentijnvdBeek committed:
Merge pull request #82 from rellermeyer/cleanup

Cleanup

ValentijnvdBeek committed:
Remove dependency on Make entirely

ValentijnvdBeek committed:
Merge branch 'cleanup' into elf_loader

ValentijnvdBeek committed:
Merge cleanup

ValentijnvdBeek commented:
@fayalalebrun Changed something with the early startup and the linker file which broke the build system. I haven't been able to figure it up until now, it should generate the DTB file correctly now but it just doesn't boot anymore. Typically this happens whenever the linker can't find the file.

fayalalebrun commented:
@ValentijnvdBeek Indeed, I believe the error must be in the linking. Something must be happening differently in the new Meson build file.

However, I'm not able to run the kernel with the Meson build system, not even get qemu to start. Could you update the README? It still refers to Make as the build system.

ValentijnvdBeek commented:
> @ValentijnvdBeek Indeed, I believe the error must be in the linking. Something must be happening differently in the new Meson build file.

Yeah, I suppose it is with the DTB file. Which one needs to be linked in? The one generated by the build file or in src/common? 

> However, I'm not able to run the kernel with the Meson build system, not even get qemu to start. Could you update the README? It still refers to Make as the build system.

Qemu not starting is typical for a bad linker script, but I cannot tell what is wrong what is wrong with it now.

ValentijnvdBeek committed:
Update README and move directories

fayalalebrun commented:
> Yeah, I suppose it is with the DTB file. Which one needs to be linked in? The one generated by the build file or in src/common?
The file in src/common is a binary file, so you can't link it directly. Hence an object file is generated from it by the build script, which is the one you have to link.

fayalalebrun commented:
When I run it, I get `arm-none-eabi-gcc: error: dtb.o: No such file or directory`

Perhaps kernel_elf is not set to depend on the dtb generation?

ValentijnvdBeek committed:
Fix dtb issue; Remove debugging calls; Fix broken merge stuff

ValentijnvdBeek committed:
Fix the build system (hopefully) #2

ValentijnvdBeek committed:
Small CI issue
